py/vtdb/vtgatev2.py

```diff
@@ -298,60 +298,91 @@ def _execute_entity_ids(
+        self._update_session(response)
+        if response.reply.get('Error'):
+          raise gorpc.AppError(response.reply['Error'], exec_method)
+        for reply in response.reply['List']:
+          fields = []
+          conversions = []
+          results = []
+          rowcount = 0
+          for field in reply['Fields']:
+            fields.append((field['Name'], field['Type']))
+            conversions.append(field_types.conversions.get(field['Type']))
+          for row in reply['Rows']:
+            results.append(tuple(_make_row(row, conversions)))
+          rowcount = reply['RowsAffected']
+          lastrowid = reply['InsertId']
+          rowsets.append((results, rowcount, lastrowid, fields))
+      except gorpc.GoRpcError as e:
+        self.logger_object.log_private_data(bind_variables_list)
+        raise self._convert_exception(
+            e, sql_list, exec_method,
+            keyspace='', tablet_type=tablet_type)
+      except Exception:
+        logging.exception('gorpc low-level error')
+        raise
+      return rowsets
 
+    # Return merged keyspace_ids and shards rowsets.
     rowsets = []
-
-    try:
-      req = {
-          'Queries': query_list,
-          'TabletType': tablet_type,
-          'AsTransaction': as_transaction,
-      }
-      self._add_caller_id(req, effective_caller_id)
-      self._add_session(req)
-      response = self._get_client().call('VTGate.ExecuteBatchKeyspaceIds', req)
-      self._update_session(response)
-      if response.reply.get('Error'):
-        raise gorpc.AppError(
-            response.reply['Error'], 'VTGate.ExecuteBatchKeyspaceIds')
-      for reply in response.reply['List']:
-        fields = []
-        conversions = []
-        results = []
-        rowcount = 0
-
-        for field in reply['Fields']:
-          fields.append((field['Name'], field['Type']))
-          conversions.append(field_types.conversions.get(field['Type']))
-
-        for row in reply['Rows']:
-          results.append(tuple(_make_row(row, conversions)))
-
-        rowcount = reply['RowsAffected']
-        lastrowid = reply['InsertId']
-        rowsets.append((results, rowcount, lastrowid, fields))
-    except gorpc.GoRpcError as e:
-      self.logger_object.log_private_data(bind_variables_list)
-      raise self._convert_exception(
-          e, sql_list, keyspace_ids_list,
-          keyspace='', tablet_type=tablet_type)
-    except Exception:
-      logging.exception('gorpc low-level error')
-      raise
+    keyspace_ids_rowsets = make_execute_batch_call(
+        keyspace_ids_query_list, is_keyspace_ids=True)
+    shards_rowsets = make_execute_batch_call(
+        shards_query_list, is_keyspace_ids=False)
+    keyspace_ids_iter = iter(keyspace_ids_rowsets)
+    shards_iter = iter(shards_rowsets)
+    for is_keyspace_ids in is_keyspace_ids_list:
+      if is_keyspace_ids:
+        rowsets.append(keyspace_ids_iter.next())
+      else:
+        rowsets.append(shards_iter.next())
```

Comment:
VTGate doesn't support mixing of shard and keyspace id queries. So, that flexibility on the client side is not worth it.
I guess this can be supported elegantly only in V3.